paddle/fluid/operators/mul_op.h

```diff
@@ -109,5 +110,105 @@ class MulGradKernel : public framework::OpKernel<T> {
+    dout_mat.Resize({m, n});
+
+    auto* ddx = ctx.Input<framework::LoDTensor>("DDX");
+    auto* ddy = ctx.Input<framework::LoDTensor>("DDY");
+
+    auto* dx = ctx.Output<framework::LoDTensor>("DX");
+    auto* dy = ctx.Output<framework::LoDTensor>("DY");
+    auto* ddout = ctx.Output<framework::LoDTensor>("DDOut");
+
+    if (dx != nullptr) {
+      dx->set_lod(x->lod());
+    }
+    if (dy != nullptr) {
+      dy->set_lod(y->lod());
+    }
+    if (ddout != nullptr) {
+      ddout->set_lod(dout->lod());
+    }
+
+    // allocate and reshape dx
+    dx->mutable_data<T>(ctx.GetPlace());
+    Tensor dx_mat = dx->dims().size() > 2
+                        ? framework::ReshapeToMatrix(*dx, x_num_col_dims)
+                        : *dx;
+
+    // allocate and reshape dy
+    dy->mutable_data<T>(ctx.GetPlace());
+    Tensor dy_mat = dy->dims().size() > 2
+                        ? framework::ReshapeToMatrix(*dy, y_num_col_dims)
+                        : *dy;
+
+    // allocate and reshape ddout
+    T* ddout_data = ddout->mutable_data<T>(ctx.GetPlace());
+    Tensor ddout_mat;
+    ddout_mat.ShareDataWith(*ddout);
+    ddout_mat.Resize({m, n});
+
+    // allocate ddout_mat_tmp for VADD
+    Tensor ddout_mat_tmp;
+    T* ddout_tmp_data = ddout_mat_tmp.mutable_data<T>({m, n}, ctx.GetPlace());
+
+    // init output as 0
+    math::SetConstant<DeviceContext, T> set_zero;
+    set_zero(ctx.template device_context<DeviceContext>(), &dx_mat,
+             static_cast<T>(0));
+    set_zero(ctx.template device_context<DeviceContext>(), &dy_mat,
+             static_cast<T>(0));
+    set_zero(ctx.template device_context<DeviceContext>(), &ddout_mat,
+             static_cast<T>(0));
+    set_zero(ctx.template device_context<DeviceContext>(), &ddout_mat_tmp,
+             static_cast<T>(0));
+
+    auto& dev_ctx = ctx.template device_context<DeviceContext>();
+    auto blas = math::GetBlas<DeviceContext, T>(dev_ctx);
+    if (ddx) {
+      auto ddx_mat = ddx->dims().size() > 2
+                         ? framework::ReshapeToMatrix(*ddx, x_num_col_dims)
+                         : static_cast<const Tensor&>(*ddx);
+
+      // dy = ddx' * dout. dy : K x M, ddx' : K x M, dout : M x N
+      blas.MatMul(ddx_mat, true, dout_mat, false, &dy_mat);
+      // ddout1 = ddx * y. ddx : M x K, y : K x N, ddout1 : M x N
+      blas.MatMul(ddx_mat, false, y_mat, false, &ddout_mat);
+    }
+    if (ddy) {
+      auto ddy_mat = ddy->dims().size() > 2
+                         ? framework::ReshapeToMatrix(*ddy, y_num_col_dims)
+                         : static_cast<const Tensor&>(*ddy);
+      // dx = dout * ddy'. dout : M x N, ddy' : N x K, dx : M x K
+      blas.MatMul(dout_mat, false, ddy_mat, true, &dx_mat);
+      // ddout2 = x * ddy. x : M x K, ddy : K x N, ddout2 : M x N
+      blas.MatMul(x_mat, false, ddy_mat, false, &ddout_mat_tmp);
+    }
+
+    // ddout = ddout1 + ddout2 = ddx * y + x * ddy
+    blas.AXPY(m * n, 1.0, ddout_tmp_data, ddout_data);
```

Comment:
To add `x * ddy` to ddout ,  set the `beta` 1.0 in `MatMul`.
There is no need to use `tmp_data` and `AXPY`

Reply:
Done. Thanks